### PR TITLE
1852 implement draft and moderation flow for competitions

### LINKED ISSUE #1852: Implement Draft/Moderation Flow for Competitive Events
## Overview
Implement a draft and moderation system for Competitive Events that allows providers to create and edit events, with a moderation step before publication. The system should follow a similar pattern to the existing Workshop draft system, with appropriate adaptations for Competitive Events. The state machine details are described in #1522 

#### Create `CompetitiveEventDraft` Entity
- Create a new entity `CompetitiveEventDraft` with similar structure to `WorkshopDraft`
- Include properties:
  - `Id` (Guid)
  - `CompetitiveEventId` (Guid, nullable)
  - `ProviderId` (Guid)
  - `DraftStatus` (enum)
  - `RejectionMessage` (string)
  - `Version` (byte[])
  - `CoverImageId` (string)
  - `CompetitiveEventDraftContent` (JSON)

#### Create `CompetitiveEventDraftContent`
- Create a class to store draft content as JSON
- Include all fields from `CompetitiveEvent` that can be modified
- Add validation attributes

#### Update CompetitiveEvent States
- Add new states to `CompetitiveEventStates` enum:
  - `Draft`
  - `PendingModeration`
  - `Active`
  - `Inactive`
  - `Archived`

#### Create CompetitiveEventDraftService
- Implement CRUD operations for drafts
- Implement moderation workflow methods:
  - `Create`
  - `Update`
  - `Delete`
  - `SendForModeration`
  - `Approve`
  - `Reject`
  - `GetByProviderId`
  - `GetById`

#### Implement Image Handling
- Add support for cover image and multiple images
- Implement image upload/delete operations
- Handle image moderation

### API Endpoints (example)

Create new endpoints in `CompetitiveEventController`:

```csharp
[HttpPost]
public async Task<IActionResult> CreateDraft([FromForm] CompetitiveEventDraftDto dto)

[HttpPut("{id}")]
public async Task<IActionResult> UpdateDraft(Guid id, [FromForm] CompetitiveEventDraftUpdateDto dto)

[HttpDelete("{id}")]
public async Task<IActionResult> DeleteDraft(Guid id)

[HttpPut("{id}/send-for-moderation")]
public async Task<IActionResult> SendForModeration(Guid id)

[HttpPut("{id}/approve")]
public async Task<IActionResult> Approve(Guid id)

[HttpPut("{id}/reject")]
public async Task<IActionResult> Reject(Guid id, [FromBody] CompetitiveEventDraftRejectionDto dto)

[HttpGet("provider/{providerId}/drafts")]
public async Task<IActionResult> GetByProviderId(Guid providerId, [FromQuery] ExcludeIdFilter filter)

[HttpGet("drafts/{id}")]
public async Task<IActionResult> GetDraftById(Guid id)
```

#### Create Draft State
- Implement draft state management using NgXs
- Create actions, reducers, and effects for draft operations
- Implement draft caching mechanism

#### Create Draft List Component
- Display list of drafts for provider
- Show draft status and basic information
- Implement draft actions (edit, delete, send for moderation)

#### Create Draft Edit Component
- Implement form for editing competitive event details
- Add image upload functionality
- Implement draft saving mechanism
- Add validation

#### Create Draft Preview Component
- Show how the event will look when published
- Display all event details
- Show moderation status

#### Create `CompetitiveEventDraftService` (example)
```typescript
export class CompetitiveEventDraftService {
  createDraft(dto: CompetitiveEventDraftDto): Observable<CompetitiveEventDraftResultDto>
  updateDraft(id: string, dto: CompetitiveEventDraftUpdateDto): Observable<CompetitiveEventDraftResultDto>
  deleteDraft(id: string): Observable<void>
  sendForModeration(id: string): Observable<void>
  getDraftById(id: string): Observable<CompetitiveEventDraftResponseDto>
  getDraftsByProviderId(providerId: string, filter: ExcludeIdFilter): Observable<SearchResult<CompetitiveEventDraftViewCardDto>>
}
```

### Interfaces (example)

```typescript
interface CompetitiveEventDraftDto {
  id?: string;
  competitiveEventId?: string;
  providerId: string;
  title: string;
  shortTitle: string;
  description: string;
  // ... other fields
}

interface CompetitiveEventDraftResultDto {
  draft: CompetitiveEventDraftResponseDto;
  uploadingCoverImgResult?: OperationResult;
  uploadingImagesResults?: MultipleKeyValueOperationResult;
}

interface CompetitiveEventDraftResponseDto {
  draftId: string;
  draftStatus: CompetitiveEventDraftStatus;
  rejectionMessage?: string;
  competitiveEventDetails: CompetitiveEventDto;
}
```

## Business Rules

1. Draft Status Flow:
   - New drafts start in `Draft` status
   - Can be sent to `PendingModeration`
   - Moderator can approve (becomes `Active`) or reject (returns to `Draft`)
   - Active events can be edited (creates new draft)
   - Active events can be archived

2. Validation Rules:
   - All mandatory fields must be filled before sending for moderation
   - Images must meet size and format requirements
   - Dates must be valid and in correct order
   - Contact information must be valid

3. Access Control:
   - Only provider employees can create/edit drafts
   - Only moderators and admins can approve/reject drafts
   - Providers can only see their own drafts
   - Moderators can see all drafts

4. Draft Management:
   - Drafts can be saved and edited multiple times
   - Drafts can be deleted if not yet published
   - Published events cannot be deleted
   - Draft changes don't affect published events until approved

### PR DESCRIPTION
**Pull request by Kornienko11004**
✨ Added Competitive Event Draft Functionality
Summary
This PR introduces full support for managing Competitive Event Drafts, including:
-Domain models
-Configuration and mapping
-Service methods (CRUD + moderation)
-Controller endpoints
-Unit and integration tests

🚀 What's New
✅ Models & Configuration
-Added CompetitiveEventDraft and CompetitiveEventDraftContent entities
-Configured entity mapping and relationships
-Fixed image configuration for drafts

🔧 Services
-Implemented Create, Update, Delete, SendForModeration methods in CompetitiveEventDraftService
-Added retrieval methods: GetByProviderId, GetCompetitiveEventDraftByIdMapped
-Setup service layer for moderation support

🌐 API Endpoints
Added endpoints for:
-Creating a draft
-Updating a draft
-Deleting a draft
-Sending a draft for moderation
-Getting drafts by provider or ID
-Removing images from coverage and main content images

🧪 Tests
Added unit tests for all main service methods
Added endpoint tests for:
-CreateDraft
-UpdateDraft
-DeleteDraft
Fixed and expanded test coverage for edge cases

🧹 Refactoring & Fixes
-Refactored routes for endpoints
-Improved image upload handling
-Small fixes to models and mappings
-General service cleanup and minor enhancements

## Summary by CodeRabbit

* **New Features**
  * Introduced full support for competitive event drafts, including creation, updating, deletion, moderation (approve/reject), and detailed draft management endpoints.
  * Added new draft-related DTOs, status enums, and result objects for comprehensive draft lifecycle handling.
  * Implemented permissions and status tracking for draft moderation and approval processes.
  * Enabled image upload and management for competitive event drafts.
  * Added detailed mapping and conversion methods between draft models and DTOs.
  * Extended competitive event states and permissions to support draft workflows.
  * Integrated draft support into data context, repositories, and provider entities.
  * Added string representation override for contact information for improved display.

* **Bug Fixes**
  * None.

* **Tests**
  * Added extensive unit tests for draft-related controller endpoints and the draft service, covering success, error, and edge cases.
  * Added integration tests for the competitive event draft repository using an in-memory database.
  * Introduced test data generators for competitive event DTOs to facilitate testing.

* **Chores**
  * Registered new services, repositories, and configurations to support competitive event drafts in the application startup.
  * Enhanced database context and entity configurations to handle draft content as JSON.
  * Updated concurrency handling in repositories for draft entities.
  * Cleaned up and reorganized using directives in test files for clarity.
  * Extended image synchronization to include competitive event draft images and cover images.

### CODE REVIEW

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/ContactInfo/ContactsDto.cs [hunk 1]
```diff
@@ -25,6 +25,30 @@ public sealed class ContactsDto : IContentComparable<Contacts>, IEquatable<Conta
     [ModelBinder(BinderType = typeof(JsonModelBinder))]
     public List<SocialNetworkDto> SocialNetworks { get; set; } = [];
 
+    public override string ToString()
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Models/ContactInfo/ContactsDto.cs:28]: do we really need toString ?:)

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+        logger.LogDebug("Competitive event updating process started. CompetitiveEvent Id = {Id}.", competitiveEventV2Dto.Id);
+
+        var existingCompetitiveEvent = await competitiveEventService.GetById(competitiveEventV2Dto.Id);
+
+        if (existingCompetitiveEvent == null)
+        {
+            throw new InvalidOperationException($"There is no CompetitiveEvent with such Id. CompetitiveEvent can`t be updated.");
+        }
+
+        var draft = await competitiveEventDraftRepository.Get(whereExpression: ced => ced.Id == competitiveEventV2Dto.Id)
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: I agree with CodeRabbit that this needs checking, as draft id is not the same as real entity id.

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+
+        if (AreModeratedFieldsChanged(competitiveEventV2Dto, existingCompetitiveEvent))
+        {
+            logger.LogDebug("Moderated fields was changed. CompetitiveEvent draft creation initiated. CompetitiveEvent Id = {Id}.", competitiveEventV2Dto.Id);
+            return (await Create(competitiveEventV2Dto)).CompetitiveEventDraft.CompetitiveEventDetails;
+        }
+
+        logger.LogDebug("Moderated fields was not changed. CompetitiveEvent update initiated. CompetitiveEvent Id = {Id}.", competitiveEventV2Dto.Id);
+
+        return competitiveEventService.UpdateV2(competitiveEventV2Dto.ToDraft().ToV2CreateRequestDto()).Result.CompetitiveEventV2;
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: also agree with code rabbit, that it is better to use await here. might add an extra line of code, but would be safer

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+        }
+
+        var stringFieldsToCompare = new List<Func<CompetitiveEventDto, string>>
+        {
+            w => w.ShortTitle,
+            w => w.Title,
+            w => w.AdditionalDescription,
+            w => w.DescriptionOfTheEnrollmentProcedure,
+            w => string.Join(" | ", w.Contacts.Select(c => c.ToString())),
+            w => w.Title
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: same :) double title comparison. Maybe there should be some other field?

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs [hunk 5]
```diff
@@ -249,10 +514,155 @@ private CompetitiveEventResponseDto CreateUpdateResponse(CompetitiveEventResultD
...
+    /// <returns>Returns <see cref="CompetitiveEventDraftResponseDto"/> with the cover image removed if successful.</returns>
+    /// <response code="200">Cover image deleted successfully.</response>
+    /// <response code="400">The draft does not have a cover image.</response>
+    /// <response code="401">The user is not authenticated.</response>
+    /// <response code="403">The user does not have permission to perform this action.</response>
+    /// <response code="404">The specified draft was not found.</response>
+    /// <response code="409">The draft is not editable.</response>
+    /// <response code="500">An unexpected error occurred.</response>
+    [HttpDelete("/api/v{version:apiVersion}/competitions-drafts/{draftId}/cover-image")]
+    [HasPermission(Permissions.CompetitiveEventEdit)]
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs]: I agree with rabbit. Only people with approve rights should be able to edit drafts other than people who belong to provider. And we have `CompetitiveEventApprove = 154` permission

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs [hunk 6]
```diff
@@ -249,10 +514,155 @@ private CompetitiveEventResponseDto CreateUpdateResponse(CompetitiveEventResultD
...
+    /// <returns>Returns <see cref="CompetitiveEventDraftResponseDto"/> with multiple images removed if successful.</returns>
+    /// <response code="200">Images deleted successfully.</response>
+    /// <response code="400">No image IDs were provided.</response>
+    /// <response code="401">The user is not authenticated.</response>
+    /// <response code="403">The user does not have permission to perform this action.</response>
+    /// <response code="404">Draft not found or images not present in draft.</response>
+    /// <response code="409">The draft is not editable.</response>
+    /// <response code="500">An unexpected error occurred.</response>
+    [HttpDelete("/api/v{version:apiVersion}/competitions-drafts/{draftId}/images")]
+    [HasPermission(Permissions.CompetitiveEventEdit)]
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs]: same for `CompetitiveEventApprove = 154`

**Review by DmyMi — CHANGES_REQUESTED** [review]

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs (hunk 2 above)
Comment (Kornienko11004) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: changed to draft id insteaad of entitys id

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs (hunk 4 above)
Comment (Kornienko11004) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: no, title should be once) removed extra field for comparson

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs (hunk 3 above)
Comment (Kornienko11004) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: change as mentioned in code rabbit comment

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs (hunk 5 above)
Comment (Kornienko11004) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs]: change to permission 154

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs (hunk 6 above)
Comment (Kornienko11004) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs]: changed to 154 permission

**Review by DmyMi — APPROVED** [review]
